apply_license.py

```diff
@@ -76,76 +75,79 @@ def match_patterns(s, patterns):
+        firstline = contents.partition('\n')[0]
+        contents = contents.replace(firstline, "")
+        license = firstline + "\n" + license
+    return license  + "\n\n" + contents
 
+def check_ignored(path):
+    return subprocess.call(['git', 'check-ignore', '-q', path]) == 0
 
 
-for subdir, dirs, files in os.walk("."):
-    # Ignore tmp and build folders
-    if subdir.startswith("./tmp") or "build" in subdir:
-        continue
+
+dryrun = False
+
+for root, dirs, files in os.walk(".", topdown=True):
+    # exclude dirs to avoid traversing them at all
+    # with topdown set to True we can make in place modifications of dirs to
+    # have os.walk() skip directories
+    dirs[:] = [ d for d in dirs if not skip_path(root + os.sep + d) and not check_ignored(root + os.sep + d) ]
+
     for file in files:
-        filepath = subdir + os.sep + file
-        ext = os.path.splitext(filepath)[1]
-        if ext not in extensions:
-            continue
+        filepath = root + os.sep + file
 
-        comment = get_comment_style_for_file(filepath)
-        if comment is None:
+        # skip ignored files
+        if skip_filename(filepath) or check_ignored(filepath):
             continue
 
-        should_include = force_include(filepath)
-        if not should_include and skip_filename(filepath, file):
-            print("Excluded " + filepath)
+        license = get_license_for_file(filepath)
+        if not license:
             continue
 
         with open(filepath, "rb+") as f:
             contents = f.read()
-            # Already applied the Defold License?
-            if "Defold Foundation" in contents:
-                print("Already applied Defold license to " + filepath)
-                continue;
 
             # Some other license in the file
-            if not should_include and has_other_license(contents):
-                print("Other license in " + filepath)
+            if has_other_license(contents):
                 continue;
 
-            # Make the license text into a language specific comment, based on extension
-            lines = LICENSE.split("\n")
-            lines = [comment + line for line in lines]
-            license = "\n".join(lines)
-
-            # Preserve shebang
-            if contents.startswith("#!"):
-                firstline = contents.partition('\n')[0]
-                contents = contents.replace(firstline, "")
-                license = firstline + "\n" + license
-
-            print("Applying Defold License to " + filepath + ("  (force included)" if should_include else ""))
-            f.seek(0)
-            f.write(license + "\n\n")
-            f.write(contents)
-            f.truncate()
+            # Already applied the Defold License?
+            # Remove it and reapply if file has been modified after license year
+            if has_defold_license(contents):
+                modified_year = int(subprocess.check_output(['git', 'log', '-1', '--pretty="%ci"', filepath])[1:5])
```

Comment:
Use git log to get date of modification. This is more reliable and accurate than checking file modification date.

Reply:
Why this optional thing?
Shouldn't the copyright year be updated regardless?

Reply:
There are many different ways of writing the copyright notice. It seems to be less important nowadays, since the US has adopted the "Berne Convention for the Protection of Literary and Artistic Works".

Some only update the copyright notice when a file has been changed. Some use a span, from the first creation until most recent change.

I've now changed to a more explicit copyright notice:

```
-- Copyright 2020-2022 The Defold Foundation
-- Copyright 2014-2020 King
-- Copyright 2009-2014 Ragnar Svensson, Christian Murray
-- Licensed under the Defold License version 1.0 (the "License"); you may not use
...
```